### PR TITLE
Add compatibly for Philips scanners

### PR DESCRIPTION
**Pull request by po09i**
## Description
Shimming Toolbox has been tested mainly with Siemens data. This PR aims to test some of the features with Philips data as well. More specifically, this PR

- Adds scaling of Philips field map data
  - The scaling from the used dataset seems to be automatically rescaled by dcm2niix to [-pi, pi]. ST ensures that the input is indeed [-pi, pi].
 - Adds manufacturer constraint information in the toolbox, this means that the scanner constraints' file is optional but takes precedence over the Toolbox's data. We need to update the plugin to default to nothing.

Problems:
- The `SliceTiming` tag used to auto detect the slice ordering is not present in the BIDS file. This is a problem with Philips' Dicoms, [not dcm2niix](https://github.com/rordenlab/dcm2niix/tree/master/Philips#missing-information) since the information is not present in the DICOM.
  - Result: We need to specify the slice ordering.
- The `PhaseEncodingDirection` tag is not present in the BIDS JSON sidecar due to [missing information in the DICOM](https://github.com/rordenlab/dcm2niix/tree/master/Philips#missing-information). Instead, only `PhaseEncodingAxis` is available.
  - Result: This tag was used to understand the gradient coordinate system. This is only used for real-time shimming right now to output the gradients with the right polarity. We need to find another way to get that information.
- The `ShimSetting` tag is [not available](https://github.com/rordenlab/dcm2niix/blob/master/BIDS/README.md#manufacturer-philips). Unsure if this is a dcm2niix issue or Philips' DICOM issue.
  - Result: The bounds for the scanner shim system might not be respected since we do not know what they are for a given acquisition.

### CODE REVIEW

File: shimmingtoolbox/cli/b0shim.py [hunk 1]
```diff
@@ -955,18 +1004,28 @@ def _save_nii_to_new_dir(list_fname, path_output):
         nib.save(nii, fname_to_save)
 
 
-def _load_output_options(json_anat, fatsat):
-    options = {'fatsat': False}
+def _get_fatsat_option(json_anat, fatsat):
+    """ Return if the fat saturation option should be turned on or off. This function maily exists to resolve the 'auto'
```
Comment (po09i) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
    """ Return if the fat saturation option should be turned on or off. This function mainly exists to resolve the 'auto'
```

File: shimmingtoolbox/cli/b0shim.py [hunk 2]
```diff
@@ -944,6 +921,78 @@ def _initial_in_bounds(coefs, bounds):
...
+        absolute bounds
+
+    Args:
+        constraints (dict): Constraints of the scanner coils
+        scanner_shim_settings (dict): Dictionary containing the shim settings of the scanner ('f0', 'order1', 'order2')
+        order (int): Order of the scanner coils (0 or 1 or 2)
+        manufacturer (str): Name of the MRI manufacturer
+
+    Returns:
+        dict: Updated constraint of the scanner
```
Comment (po09i) [inline shimmingtoolbox/cli/b0shim.py]:
```suggestion
        dict: Updated constraints of the scanner
```

File: shimmingtoolbox/cli/b0shim.py [hunk 3]
```diff
@@ -325,6 +322,7 @@ def dynamic(fname_fmap, fname_anat, fname_mask_anat, method, opt_criteria, slice
 
                 # If the output format is absolute, add the initial coefs
                 if output_value_format == 'absolute':
+                    initial_coefs = scanner_shim_settings.concatenate_shim_settings(scanner_coil_order)
```
Comment (4rnaudB) [inline shimmingtoolbox/cli/b0shim.py:325]: intial_coefs is replaced by `options['scanner_shim']` in 
`list_coils = _load_coils(coils, scanner_coil_order, fname_sph_constr, nii_fmap, options['scanner_shim'],
                             json_fm_data['Manufacturer'], json_fm_data['ManufacturersModelName'])`:248
and by `scanner_shim_settings.concatenate_shim_settings(scanner_coil_order)` here. Are they the same thing?

File: shimmingtoolbox/cli/b0shim.py [hunk 4]
```diff
@@ -758,6 +756,7 @@ def realtime_dynamic(fname_fmap, fname_anat, fname_mask_anat_static, fname_mask_
                 #             bounds=bounds_shim_cs)
                 # If the output format is absolute, add the initial coefs
                 if output_value_format == 'absolute':
+                    initial_coefs = scanner_shim_settings.concatenate_shim_settings(scanner_coil_order)
```
Comment (4rnaudB) [inline shimmingtoolbox/cli/b0shim.py:759]: Same as previous comment

File: shimmingtoolbox/coils/coil.py [hunk 5]
```diff
@@ -165,32 +148,76 @@ def _create_coil_profile(self, dim):
...
+            constraints["coef_channel_minmax"].append([123100100, 123265000])
+        if order >= 1:
+            for _ in range(3):
+                constraints["coef_channel_minmax"].append([-2300, 2300])
+        if order >= 2:
+            constraints["coef_channel_minmax"].append([-4959.01, 4959.01])
+            constraints["coef_channel_minmax"].append([-3551.29, 3551.29])
+            constraints["coef_channel_minmax"].append([-3503.299, 3503.299])
+            constraints["coef_channel_minmax"].append([-3551.29, 3551.29])
+            constraints["coef_channel_minmax"].append([-3487.302, 3487.302])
```
Comment (4rnaudB) [inline shimmingtoolbox/coils/coil.py]: Could be 1 line using `.extend([[-4959.01, 4959.01], [-3551.29, 3551.29], ...])`

File: shimmingtoolbox/cli/b0shim.py [hunk 6]
```diff
@@ -944,6 +921,78 @@ def _initial_in_bounds(coefs, bounds):
...
+                    logger.warning(f"Initial scanner coefs are outside the bounds allowed in the constraints: "
+                                   f"{bounds[i_bound]}, initial: {coefs[i_bound]}")
+            if bounds[i_bound][0] is not None:
+                if not bounds[i_bound][0] <= coefs[i_bound]:
+                    logger.warning(f"Initial scanner coefs are outside the bounds allowed in the constraints: "
+                                   f"{bounds[i_bound]}, initial: {coefs[i_bound]}")
+
+    if order == 0:
+        initial_coefs = [0]
+    elif order == 1:
```
Comment (Aurelienpujolm) [inline shimmingtoolbox/cli/b0shim.py:957]: Can you explain, why you are doing this here ? (maybe add a quick comment to explain where these numbers came from)

File: shimmingtoolbox/shim/sequencer.py [hunk 7]
```diff
@@ -1463,7 +1463,7 @@ def new_bounds_from_currents(currents, old_bounds):
 
     Args:
         currents (np.ndarray): 2D array (n_shims x n_channels). Direct output from :func:`_optimize`.
```
Comment (Aurelienpujolm) [inline shimmingtoolbox/shim/sequencer.py]: Maybe specify, in which class _optimize is ?   ( Sequencer)

File: shimmingtoolbox/shim/shim_utils.py [hunk 8]
```diff
@@ -191,3 +191,114 @@ def shim_to_phys_cs(coefs, manufacturer):
...
+                scanner_shim_mp['order1'] = order1_mp
+
+        if shim_settings.get('order2'):
+            max_current_mp_order2 = np.array([4959.01, 3551.29, 3503.299, 3551.29, 3487.302])
+            max_current_dcm_order2 = np.array([9998, 9998, 9998, 9998, 9998])
+            order2_mp = np.array(shim_settings['order2']) * max_current_mp_order2 / max_current_dcm_order2
+
+            if np.any(np.abs(order2_mp) > max_current_mp_order2):
+                scanner_shim_mp['has_valid_settings'] = False
+                raise ValueError("Multipole values exceed known system limits.")
```
Comment (Aurelienpujolm) [inline shimmingtoolbox/shim/shim_utils.py:275]: typo mistake + why are you saying multiples, in some case, there can be, just one value that is not respecting limits

**Review by Aurelienpujolm — COMMENTED** [review]
It's working for me, but there is some points that are not clear for me, you can maybe add some more comments !

File: shimmingtoolbox/cli/b0shim.py (hunk 3 above)
Comment (po09i) [reply shimmingtoolbox/cli/b0shim.py:325]: They used to be the same thing, now they represent the same thing but are different.

File: shimmingtoolbox/cli/b0shim.py (hunk 4 above)
Comment (po09i) [reply shimmingtoolbox/cli/b0shim.py:759]: See above message

File: shimmingtoolbox/cli/b0shim.py (hunk 3 above)
Comment (po09i) [reply shimmingtoolbox/cli/b0shim.py:325]: the concatenate method makes sure it is the same format as it was before.

File: shimmingtoolbox/shim/shim_utils.py (hunk 8 above)
Comment (po09i) [reply shimmingtoolbox/shim/shim_utils.py:275]: This is not a typo. Mulipole is the type of unit

File: shimmingtoolbox/coils/coil.py (hunk 5 above)
Comment (po09i) [reply shimmingtoolbox/coils/coil.py]: Updated: [86e0c38](https://github.com/shimming-toolbox/shimming-toolbox/pull/435/commits/86e0c38edb9bdde394f35afdfb714d4f418c6f78)

File: shimmingtoolbox/cli/b0shim.py (hunk 6 above)
Comment (po09i) [reply shimmingtoolbox/cli/b0shim.py:957]: Provided more info here: [86e0c38](https://github.com/shimming-toolbox/shimming-toolbox/pull/435/commits/86e0c38edb9bdde394f35afdfb714d4f418c6f78)

File: shimmingtoolbox/shim/sequencer.py (hunk 7 above)
Comment (po09i) [reply shimmingtoolbox/shim/sequencer.py]: I removed the comment as it is now a class and did not provide much more information

**Review by 4rnaudB — APPROVED** [review]

**Pull request merged**